### PR TITLE
Introduce proxies as a means of breaking circular import chains

### PR DESCRIPTION
**Pull request by skizzerz**
What it says on the tin

### CODE REVIEW

**Comment by skizzerz** [comment]
Don't merge yet: should support multiple proxy stubs for a single implementation, provided all the stubs have the same signature (e.g. gamemodes.py can declare a stub for something, and roles.py can declare a stub for something, and wolfgame.py can define the implementation of that and it should work, provided the two stubs share the same signature).

File: src/proxy.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import inspect
+
+from src.decorators import handle_error
```
Comment (Vgr255) [inline src/proxy.py]: Seems to be an artefact; it's not used.

File: src/proxy.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        SIGS[f.__name__] = inspect.signature(f)
+    else:
+        # Python 3.2; not allowed to have nice things
+        SIGS[f.__name__] = None
+
+    return inner
+
+def impl(f):
+    if f.__name__ not in SIGS:
+        raise NameError(("Attempting to implement a proxy stub {0} that does "
```
Comment (Vgr255) [inline src/proxy.py]: Nit: Use `{0!r}` to have quotes around the name; makes it more readable imo.

File: src/wolfgame.py [hunk 3]
```diff
@@ -2074,6 +2067,7 @@ def fday(cli, nick, chan, rest):
 
 # Specify force = "nick" to force nick to be lynched
 @handle_error
```
Comment (Vgr255) [inline src/wolfgame.py]: Can remove that, `impl` already wraps.

File: src/proxy.py (hunk 2 above)
Comment (Vgr255) [inline src/proxy.py]: Also bunch of other places where this'd nicer-looking.

**Comment by Vgr255** [comment]
Couple of nitpicks pointed out, other than that looks fine.

File: src/wolfgame.py (hunk 3 above)
Comment (skizzerz) [inline src/wolfgame.py]: No it doesn't. it wraps it for things called using the stub, but the original method is not wrapped in impl.

Comment (Vgr255) [inline src/wolfgame.py]: Oh, right.

**Pull request merged**